appinventor/buildserver/src/com/google/appinventor/buildserver/Compiler.java

```diff
@@ -985,22 +993,43 @@ private boolean prepareApplicationIcon(File outputPngFile) {
     return true;
   }
 
-  private boolean runDx(File classesDir, String dexedClasses) {
+  private boolean runDx(File classesDir, String dexedClassesDir, boolean secondTry) {
+    List<File> libList = new ArrayList<File>();
     List<File> inputList = new ArrayList<File>();
+    List<File> class2List = new ArrayList<File>();
     inputList.add(classesDir); //this is a directory, and won't be cached into the dex cache
     inputList.add(new File(getResource(SIMPLE_ANDROID_RUNTIME_JAR)));
     inputList.add(new File(getResource(KAWA_RUNTIME)));
     inputList.add(new File(getResource(ACRA_RUNTIME)));
 
-    // Add libraries to command line arguments
-    System.out.println("Libraries needed command line n = " + librariesNeeded.size());
     for (String library : librariesNeeded) {
-      inputList.add(new File(getResource(RUNTIME_FILES_DIR + library)));
+      libList.add(new File(getResource(RUNTIME_FILES_DIR + library)));
+    }
+
+    int offset = libList.size();
+    // Note: The choice of 12 libraries is arbitrary. We note that things
+    // worked to put all libraries into the first classes.dex file when we
+    // had 16 libraries and broke at 17. So this is a conservative number
+    // to try.
+    if (!secondTry) {           // First time through, try base + 12 libraries
+      if (offset > 12)
+        offset = 12;
+    } else {
+      offset = 0;               // Add NO libraries the second time through!
+    }
+    for (int i = 0; i < offset; i++) {
```

Comment:
I don't understand what's happening on this second try.  Is it that the first try is handling 12 libraries and if that fails, we  allow more?  Could you add a comment to explain this better?